pkg/image/skopeo.go

```diff
@@ -0,0 +1,180 @@
+	CopyImage(string, string, string, string) error
+}
+
+type skopeoOperations struct{}
+
+type manifest struct {
+	SchemaVersion int     `json:"schemaVersion"`
+	Layers        []layer `json:"layers"`   // schemaVersion v2
+	FsLayers      []layer `json:"fsLayers"` // schemaVersion v1
+}
+type layer struct {
+	Digest  string `json:"digest"`  // schemaVersion v2
+	BlobSum string `json:"blobSum"` // schemaVersion v1
+}
+
+var skopeoExecFunction = system.ExecWithLimits
+
+var processLimits = &system.ProcessLimitValues{AddressSpaceLimit: maxMemory, CPUTimeLimit: maxCPUSecs}
+
+var skopeoInterface = NewSkopeoOperations()
+
+// NewSkopeoOperations returns the default implementation of SkopeoOperations
+func NewSkopeoOperations() SkopeoOperations {
+	return &skopeoOperations{}
+}
+
+func (o *skopeoOperations) CopyImage(url, dest, accessKey, secKey string) error {
+	var err error
+	if len(accessKey) > 0 && len(secKey) > 0 {
+		var creds = "--src-creds=" + accessKey + ":" + secKey
+		_, err = skopeoExecFunction(processLimits, nil, "skopeo", "copy", url, dest, creds)
+	} else {
+		_, err = skopeoExecFunction(processLimits, nil, "skopeo", "copy", url, dest)
+	}
+	if err != nil {
+		return errors.Wrap(err, "could not copy image")
+	}
+	return nil
+}
+
+// CopyImage download image from registry with skopeo
+func CopyImage(url, dest, accessKey, secKey string) error {
+	glog.V(1).Infof("copying %q to %q...\n", url, dest)
+	skopeoDest := "dir:" + dest + dataTmpDir
+	err := skopeoInterface.CopyImage(url, skopeoDest, accessKey, secKey)
+	if err != nil {
+		return errors.Wrap(err, "Failed to download from registry")
+	}
+	err = extractImageLayers(dest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to extract image layers")
+	}
+
+	// Clean temp folder
+	os.RemoveAll(dest + dataTmpDir)
+
+	return err
+}
+
+func extractImageLayers(dest string) error {
+	glog.V(1).Infof("extracting image layers to %q\n", dest)
+	// Parse manifest file
+	manifest, err := getImageManifest(dest + dataTmpDir)
+	if err != nil {
+		return err
+	}
+
+	// Extract layers
+	var layers []layer
+	if manifest.SchemaVersion == 1 {
+		layers = manifest.FsLayers
+	} else {
+		layers = manifest.Layers
+	}
+	for _, m := range layers {
+		var layerID string
+		if manifest.SchemaVersion == 1 {
+			layerID = m.BlobSum
+		} else {
+			layerID = m.Digest
+		}
+		layer := strings.TrimPrefix(layerID, "sha256:")
+		file := fmt.Sprintf("%s%s/%s", dest, dataTmpDir, layer)
+		err := extractTar(file, dest)
+		if err == nil {
+			err = cleanWhiteoutFiles(dest)
+		}
+	}
+	return err
+}
+
+func getImageManifest(dest string) (*manifest, error) {
+	// Open Manifest.json
+	manifestFile, err := ioutil.ReadFile(dest + "/manifest.json")
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read manifest file")
+	}
+
+	// Parse json file
+	var manifestObj manifest
+	json.Unmarshal(manifestFile, &manifestObj)
+	return &manifestObj, nil
+}
+
+func extractTar(file, dest string) error {
```

Comment:
So we have several other places using tar (the cloner target, and tar extraction), the cloner target has an 'untar' function that takes a Reader and target directory parameter. Maybe we can move that into utils and re-use it for all the different places? Then you can simply open a file reader, and pass that.